File: Lab codes/Data Structure and Algorithms/Linkedlist.c
// importing libs
#include<stdio.h>
#include<stdlib.h>
#include<malloc.h>

// node declaration
struct node
{
    int data;
    struct node *link;
}*header,*ptr,*newnode,*ptr1;

// insertion at begin
void insertatBegin(int data)
{
    newnode = (struct node*)malloc(sizeof(struct node));
    newnode->data = data;
    if(header == NULL)
        newnode->link = NULL;
    else
        newnode->link = header;
    header = newnode;
}

// insertion at end

void insertatLast(int data)
{
    newnode = (struct node*)malloc(sizeof(struct node));
    newnode->data = data;
    newnode->link = NULL;
    if(header == NULL)
        header = newnode;
    else
    {
        ptr = header;
        while(ptr->link!=NULL)
            ptr = ptr->link;
        ptr->link = newnode;
    }
    
}

//insertion at position

void insertatPosition(int data,int pos)
{
    newnode = (struct node*)malloc(sizeof(struct node));
    newnode->data = data;
    ptr = header;
    for(int i=1;i<pos-1;i++)
    {
        ptr = ptr->link;
    }
    newnode->link = ptr->link;
    ptr->link = newnode;
}

// deletion at beginning

void deleteatStart()
{
    if(header == NULL)
        printf("\nList is empty");
    else
    {
        ptr = header;
        header =header->link;
        free(ptr);
    }
}

// deletion at the end

void deleteatEnd()
{
    if(header == NULL)
    {
        printf("\nList is empty");
    }
    ptr = header;
    while(ptr->link!= NULL)
    {
        ptr1 = ptr;
        ptr = ptr->link;
    }
    ptr1->link = NULL;
    free(ptr);
}

// deletion at a position

void deleteatPosition(int pos)
{
    ptr =header;
    for(int i =0;i<pos;i++)
    {
        ptr1 = ptr;
        ptr = ptr->link;
    }
    ptr1->link= ptr->link;
    free(ptr);

}

void display()
{
    if(header == NULL)
    {
        printf("\nList is empty");
    }
    else
    {
        ptr = header;
        while(ptr!=NULL)
        {
            printf("%d ",ptr->data);
            ptr = ptr->link;
        }
    }
    printf("\n");
}

void search(int key)
{
    int i =0;
    ptr = header;
    while(ptr!=NULL)
    {
        i++;
        if(ptr->data == key)
        {
            printf("Item %d is found at location: %d\n",key,i);
            break;
        }
        ptr = ptr->link;
    }

}

int main()
{
    header = NULL;
    int i,x;
    int a[]={2,9,10,5,6,3,1};
    int n = sizeof(a)/sizeof(a[0]);
    for(i=0;i<n;i++)
        insertatLast(a[i]);
    display();
    insertatLast(11);
    display();
    insertatBegin(12);
    display();
    insertatPosition(14,4);
    display();
    search(10);
    deleteatPosition(2);
    display();
    return 0;
}

File: Practice/Python/Linked List/linkedlist.py
# Linked list definition
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class LinkedList:

    def __init__(self):
        self.head = None
    #insertion statements
    #push for insertion at beginning
    #append for insertion at the end
    #insertAfter for insertion after the position

    #insertion at the beginning
    def push(self,new_data):

        new_node = Node(new_data)
        new_node.next = self.head

        self.head = new_node
    
    #insertion at the end
    def append(self,new_data):

        new_node = Node(new_data)
        if self.head is None:
            self.head = new_node
            return
        
        ptr = self.head
        while(ptr.next):
            ptr = ptr.next

        ptr.next = new_node

    #insertion after the  position
    def insertAfter(self,prev_node,new_data):
        if prev_node is None:
            print("The given previous node must in LinkedList.")
            return
        new_node = Node(new_data)
        new_node.next = prev_node.next

        prev_node.next = new_node

    #deletion statements
    #beginning
    #End
    #Middle

    #deletion at beginning
    def deleteAtBegin(self):
        if self.head is None:
            print("List is empty")
        else:
            temp = self.head
            self.head = self.head.next
            del temp
    
    #deletion at the end
    def deleteAtEnd(self):
        ptr = self.head
        while(ptr.next):
            ptr1 = ptr
            ptr=ptr.next
        ptr1.next = None
        del ptr
    
    #deletion at the position
    def deleteAtPos(self,pos:int):
        ptr = self.head
        for i in range(pos):
            ptr1 = ptr
            ptr = ptr.next
        ptr1 = ptr.next
        del ptr

    def display(self):
        ptr = self.head
        while(ptr):
            print(f"{ptr.data}",end=" ")
            ptr = ptr.next
        print()
    
    def search(self,val)->bool:
        ptr = self.head
        while(ptr):
            if(ptr.data==val):
                return True
            ptr=ptr.next
        return False
    
    

if __name__ == "__main__":
    li = LinkedList()
    for i in range(1,5):
        li.append(i)
    li.display()
    if(li.search(4)):
        print("In linked list")
    else:
        print("Not in linked list")
    print(str(li.getMiddle()))



        
        
        

